src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+    static constexpr auto fallback_fragment_source = R"(
+            #version 330 core
+            out vec4 FragColor;
+            
+            void main()
+            {
+                FragColor = vec4(255.0f, 0.0f, 255.0f, 1.0f);
+            }    
+        )";
+
+    [[nodiscard]] constexpr std::string_view ShaderProgram::get_name_from_type(Type const type) {
+        switch (type) {
+            case Type::Vertex:
+                return "VERTEX";
+            case Type::Geometry:
+                return "GEOMETRY";
+            case Type::Fragment:
+                return "FRAGMENT";
+        }
+        return "INVALID";
+    }
+
+    [[nodiscard]] GLuint ShaderProgram::compile(Type const type, std::string_view const source) {
+        // compile
+
+        auto error_message = [&](GLuint id) -> std::string {
+            auto message = std::string{};
+            auto len = GLint{};
+            glGetShaderiv(id, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow_cast<usize>(len - 2));
+            glGetShaderInfoLog(id, len - 1, nullptr, message.data());
+            return message;
+        };
+        auto compile_single = [&](GLuint& id, char const* single_source) -> GLint {
+            id = glCreateShader(sly::to_underlying(type));
+            glShaderSource(id, 1, &single_source, nullptr);
+            glCompileShader(id);
+            auto success = GLint{};
+            glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+            return success;
+        };
+
+        auto id = GLuint{};
+        auto success = compile_single(id, source.data());
+
+        if (not success) {
+
+            auto message = error_message(id);
+            spdlog::critical("ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}", get_name_from_type(type), message);
+            glDeleteShader(id);
+
+            spdlog::info("SHADER::{}::TRY_FALLBACK_COMPILATION", get_name_from_type(type));
+            switch (type) {
+                case Type::Vertex:
+                    success = compile_single(id, fallback_vertex_source);
+                    if (not success) {
+                        auto message2 = error_message(id);
+                        spdlog::critical(
+                                "ERROR::SHADER::{}::COMPILATION_FAILED\nerror: {}",
+                                get_name_from_type(type),
+                                message2
+                        );
+                        glDeleteShader(id);
+                        throw GLError(
+                                GLErrorType::FailedToCompileVertexShader,
+                                fmt::format("first error: {}\nsecond error: {}", message, message2)
+                        );
+                    }
+                    break;
+
+                case Type::Geometry:
+                    spdlog::critical("ERROR::SHADER::{}::NO_FALLBACK", get_name_from_type(type));
+                    throw GLError(GLErrorType::FailedToCompileGeometryShader, message);
```

Comment:
Do we want to throw here or is an error message enough? Maybe it's okay to ignore the error (and just log the message).